src/main/java/com/netflix/imflibrary/writerTools/IMPAssembler.java

```diff
@@ -0,0 +1,387 @@
+                            mediaFile.getRepeatCount(), // defaults to 1 if null
+                            UUIDHelper.fromUUID(UUID.randomUUID()),   // used as the essence descriptor id
+                            hash,
+                            CompositionPlaylistBuilder_2016.defaultHashAlgorithm
+                        )
+                );
+            }
+            // add to virtual tracks
+            logger.info("Creating virtual track..");
+            Composition.VirtualTrack virtualTrack = new IMFEssenceComponentVirtualTrack(
+                    UUID.randomUUID(),
+                    track.getSequenceTypeEnum(),
+                    resources,
+                    simpleTimeline.getEditRate()
+            );
+            virtualTracks.add(virtualTrack);
+        }
+
+        logger.debug("Created list of virtual tracks: {}", virtualTracks);
+        logger.debug("Created track file metadata map: {}", imfTrackFileMetadataMap);
+
+        logger.info("Building IMP here: {}...", outputDirectory.getAbsolutePath());
+        imfErrors.addAllErrors(IMPBuilder.buildIMP_2016("IMP",
+                "Netflix",
+                virtualTracks,
+                simpleTimeline.getEditRate(),
+                "http://www.smpte-ra.org/schemas/2067-21/2016",
+                imfTrackFileMetadataMap,
+                outputDirectory));
+
+        logger.info("Listing files in output dir..");
+        if (outputDirectory.isDirectory()) {
+            File[] files = outputDirectory.listFiles();
+            if (files != null) {
+                for (File file : files) {
+                    if (file != null) {
+                        logger.info("File in output dir: {}", file.getAbsolutePath());
+                    }
+                }
+            }
+        }
+
+        BasicMapProfileV2MappedFileSet mapProfileV2MappedFileSet = new BasicMapProfileV2MappedFileSet(outputDirectory);
+        AssetMap assetMap = mapProfileV2MappedFileSet.getAssetMap();
+        File assetMapOutputFile = new File(mapProfileV2MappedFileSet.getAbsoluteAssetMapURI());
+        File pklOutputFile = null;
+        File cplOutputFile = null;
+        List<File> outputMediaFiles = new ArrayList<>();
+
+
+        if (assetMap.getPackingListAssets().size() > 1) {
+            throw new IllegalArgumentException("More than one packing list found in the output asset map");
+        }
+        for (AssetMap.Asset packingListAsset: assetMap.getPackingListAssets()) {
+            if (packingListAsset.isPackingList()) {
+                pklOutputFile = new File(outputDirectory, packingListAsset.getPath().toString());
+                PackingList packingList = new PackingList(pklOutputFile);
+                for (PackingList.Asset asset : packingList.getAssets()) {
+                    logger.debug("Asset from packing list: {}", asset);
+                    // is there a better way to identify a CPL?
```

Comment:
This isn't a great way to identify the CPL, but technically it _will_ work for your specific usage.

The IMF standards do not dictate that CPL file names follow the pattern `"CPL-" + uuid + ".xml"`. That is just part of Photon's IMPBuilder implementation. Now since (internally) you are using Photon's IMPBuilder, technically, your CPL's always _should_ follow that pattern. But it is a fragile implementation, and may break if someone changes the file name generator code elsewhere.
I am also somewhat hesitant to do things that reinforce Photon's implementation as a "de-facto standard". So if we can avoid adding code that makes assumptions based on file names, I think that is the best bet.

Finding _which_ files contain the CPL requires actually _looking into_ the individual files. But you already have the PKL here, and that tells you the MIME type of each file. You can filter to only files of 'text/xml', and then just look for an XML document that has 'CompositionPlaylist' as the document element. Here is an 
[example where this logic already exists](https://github.com/Netflix/photon/blob/dd30befbfcb5dfcb3b68538d71dc9040c834108e/src/main/java/com/netflix/imflibrary/app/IMPFixer.java#L233) to find the CPL.

Reply:
Thank you for the reference, I'll use that method.